### PR TITLE
Permissions system for ships/outfits

### PR DESCRIPTION
**Pull request by a-random-lemurian**
Implements #6387
May be related to #5311

Note the `unsellable 1` attribute.

Put this in a plugin file somewhere.

You can put the Unsellable Cooling Ducts in storage, but you can't sell it!
```
outfit "Unsellable Cooling Ducts"
	unsellable 1
	plural "Cooling Ducts"
	category "Systems"
	cost 10000
	thumbnail "outfit/cooling ducts"
	"mass" 5
	"outfit space" -5
	"cooling" 5
	description "When trying to cram as many high-powered outfits as possible into a ship, occasionally you reach the point where it is sporadically overheating and shutting down. In that case, you may need to install one of these. And you can't sell this..."
```

### CODE REVIEW

File: source/Outfit.cpp [hunk 1]
```diff
@@ -196,6 +196,8 @@ void Outfit::Load(const DataNode &node)
 				if(isNewLicense(licenses, grand.Token(0)))
 					licenses.push_back(grand.Token(0));
 		}
+		else if(child.Token(0) == "unsellable")
```
Comment (Hurleveur) [inline source/Outfit.cpp:199]: I think having user write `"unsellable"` 1 for the outfit would work too and it will automatically add itself into the dictionary (in the same way we write `"automaton"` 1 for some ships)

Comment (a-random-lemurian) [reply source/Outfit.cpp]: Alright, I performed a force push.

**Comment by petervdmeer** [comment]
The original request in #6387 talks about outfits that once installed cannot be undone (permanent modifications to a ship). This suggests that uninstall should also be blocked (and that we might want to rename the tag to `persistent` or something else that reflects the "install once, never uninstall it" idea).

An `"unsellable"` item that can be uninstalled sounds not very useful to me as a story tool; the player can just put the item on another ship and then sell or disown that other ship to get rid of the item.

**Comment by petervdmeer** [comment]
One more note "un-uninstallable" also implies that the outfit is unplunderable.

**Comment by a-random-lemurian** [comment]
> The original request in #6387 talks about outfits that once installed cannot be undone (permanent modifications to a ship). This suggests that uninstall should also be blocked (and that we might want to rename the keyword to `persistent` or something else that reflects the "install once, never uninstall it" idea).
> 
> An `"unsellable"` item that can be uninstalled sounds not very useful to me as a story tool; the player can just put the item on another ship and then sell or disown that other ship to get rid of the item.

Should we make it un-uninstallable then?

I'm also thinking of adding a confirmation dialog, so before you install it there's a warning, something along the lines of:

```
Once bought, this item cannot be
sold or uninstalled. Do you want
to install this outfit?
```

**Review by tehhowch — CHANGES_REQUESTED** [review]
Note that the issue this is based off of is now closed; it duplicates an existing request. On that note, this PR's current implementation does not satisfy the needs of the required functionality (that is, it is not scalable to make managing permissions on an entity use the "attributes" field), in addition to having side effects (any ship that has an "unsellable" outfit installed will respond to an attribute query saying _the ship itself_ is "unsellable"). While that propagation could be desirable functionality to some, achieving it by accident is not.

In short: as-is, this PR won't merge. The solution we eventually implement that enables making an outfit unable to be sold if purchased will also need to be logically extendable to making a ship unsellable. It should also make sense for expanding to incorporate other outfit and ship permissions/restrictions, such as the ability to uninstall (outfits) and rename / refit (ships).

**Comment by a-random-lemurian** [comment]
> Note that the issue this is based off of is now closed; it duplicates an existing request. On that note, this PR's current implementation does not satisfy the needs of the required functionality (that is, it is not scalable to make managing permissions on an entity use the "attributes" field), in addition to having side effects (any ship that has an "unsellable" outfit installed will respond to an attribute query saying _the ship itself_ is "unsellable"). While that propagation could be desirable functionality to some, achieving it by accident is not.
> 
> In short: as-is, this PR won't merge. The solution we eventually implement that enables making an outfit unable to be sold if purchased will also need to be logically extendable to making a ship unsellable. It should also make sense for expanding to incorporate other outfit and ship permissions/restrictions, such as the ability to uninstall (outfits) and rename / refit (ships).

Is this PR salvageable, or will I need to open a separate PR?

**Comment by petervdmeer** [comment]
> Is this PR salvageable, or will I need to open a separate PR if I get around to implementing permissions for ship renaming/refitting, alongside the other changes you requested?

I feel that you don't need to implement everything that is asked in #5311 to get to a merge-able PR, as long as your implementation does follow the ideas in #5311.

So your (new or salvaged) PR should have a generic implementation of the `inability` keyword on Outfits, and one (or both) of the following combinations:

```
	inability
		uninstall
```

or 

```
	inability
		sell
```

Implementing the `inability` and `lock` keywords on ships and any of the permissions on ships should not be required to get a merge-able PR (as long as you don't implement things that conflict with such keywords on ships later).

I like your idea of giving a warning before installing an un-uninstallable outfit; and if you come up with such a good idea as PR-creator, then I also expect that to be in the PR before I consider the PR to be merge-able. :-)

**Comment by tehhowch** [comment]
Indeed, not unsalvageable.

There are still uncertainties on the syntax to use, so those should get ironed out in the parent issue before implementing. I encourage discussion of keywords and what "common" usages of the system would look like with each syntax, and whether the keyword will be intuitive when applied to both outfits and ships.

(Not all things mentioned in the issue will end up making sense to implement in the long run, either.)

**Comment by MasterOfGrey** [comment]
> An `"unsellable"` item that can be uninstalled sounds not very useful to me as a story tool; the player can just put the item on another ship and then sell or disown that other ship to get rid of the item.

Presumably the ship would acquire the unsellable attribute though no? Requiring it to be dumped/disowned/ejected instead?

**Comment by Zitchas** [comment]
I would just like to point in that having an outfit that is unsellable on a ship need not always mean that the ship is unsellable.

There are a *lot* of potential modifications I can think of doing to a ship that would be impossible to uninstall or sell; but that just means they are permanent parts of the ship. If you ever want to get rid of the ship... Well, you are free to do so. The caveat is that these "permanent" things should be destroyed when the ship is sold, and not showing up in the outfitter as stuff that is temporarily available like other outfits do.

Different types of restrictions I can think of:
1. Permanent (once part of a ship, it can never be removed from the ship)
2. Forbidden outfits (you can acquire them and install/uninstall, but no-one will touch it. Eg. Certain Quarg outfits that supposedly everyone is deathly afraid of being caught with. So outfitters won't even take them for free. You literally have to uninstall to cargo and ditch them into space)
3. Character-based (things like EVN's psychic abilities, or stuff representing the pilot's skills, attributes, etc. These should be locked to the pilot, and just go with them. In effect, they're always equipped on the flagship, and automatically transfer to the new flagship whenever the player changes ship. They can't be sold, they can't be uninstalled, they can't be ditched into space. They just are always there.)

From a content creator perspective, that `inability` keyword section looks pretty clear and usable.

**Comment by a-random-lemurian** [comment]
> I would just like to point in that having an outfit that is unsellable on a ship need not always mean that the ship is unsellable.
> 
> There are a _lot_ of potential modifications I can think of doing to a ship that would be impossible to uninstall or sell; but that just means they are permanent parts of the ship. If you ever want to get rid of the ship... Well, you are free to do so. The caveat is that these "permanent" things should be destroyed when the ship is sold, and not showing up in the outfitter as stuff that is temporarily available like other outfits do.
> 
> Different types of restrictions I can think of:
> 
>     1. Permanent (once part of a ship, it can never be removed from the ship)
> 
>     2. Forbidden outfits (you can acquire them and install/uninstall, but no-one will touch it. Eg. Certain Quarg outfits that supposedly everyone is deathly afraid of being caught with. So outfitters won't even take them for free. You literally have to uninstall to cargo and ditch them into space)
> 
>     3. Character-based (things like EVN's psychic abilities, or stuff representing the pilot's skills, attributes, etc. These should be locked to the pilot, and just go with them. In effect, they're always equipped on the flagship, and automatically transfer to the new flagship whenever the player changes ship. They can't be sold, they can't be uninstalled, they can't be ditched into space. They just are always there.)
> 
> 
> From a content creator perspective, that `inability` keyword section looks pretty clear and usable.

So in the `inability` section, what keyword would you suggest for the character-based restriction?

**Review by petervdmeer — CHANGES_REQUESTED** [review]
I like the direction in which this PR is heading.

But there are some open items which needs to be addressed before this PR is ready:
- The description of the PR needs to be updated with the actual syntax and keywords added in this PR.
- The default PR template has a section that desribes how the PR was tested, I would like to see this information also in this PR. (And preferably in the default template, because using a default template makes it easier for reviewers to more quickly review the PR.)
- The default PR template also has a section with instructions on how reviewers can quickly test the workings of the PR (usually including by a savegame or a datafile that reviewers can put in their data-directory).
- The new Permissions.cpp/h files needs to be added to XCode; not a big deal, the [CI system already generated a patch just to do that](https://github.com/endless-sky/endless-sky/suites/5216816275/artifacts/159744756).

Please also consider changing this PR to Draft if you are still working on it (so that you can explicitly set it to non-Draft once you feel the PR is ready).

**Comment by petervdmeer** [comment]
> So in the `inability` section, what keyword would you suggest for the character-based restriction?

I would suggest to keep that for a completely separate PR. My expectation is that character-based restrictions would work quite differently from ship-restrictions or outfit-restrictions.

**Comment by petervdmeer** [comment]
> 3. Character-based (things like EVN's psychic abilities, or stuff representing the pilot's skills, attributes, etc. These should be locked to the pilot, and just go with them. In effect, they're always equipped on the flagship, and automatically transfer to the new flagship whenever the player changes ship. They can't be sold, they can't be uninstalled, they can't be ditched into space. They just are always there.)

I disagree with the "they are always equipped on the flagship" part. If the player is landed, then logically those abilities are not equipped on the flagship (and I would expect such abilities to be equipped on launch and de-equipped on landing and de-equipped/re-equipped on switching ships in space).
I feel this is really out of scope for this PR (and for the referenced issues); could you file a new issue to request this feature so that we can discuss the feature and the keywords/syntax there?

File: source/OutfitterPanel.cpp [hunk 2]
```diff
@@ -579,6 +579,9 @@ bool OutfitterPanel::CanSell(bool toStorage) const
 	if(!planet || !selectedOutfit)
 		return false;
 
+	if(selectedOutfit->Get("unsellable") && !toStorage)
```
[the remark this answers is not in the document]
Comment (a-random-lemurian) [reply source/OutfitterPanel.cpp:582]: I rebased the PR branch onto a more recent master commit, but as I have said, I'm gonna fix this.

File: source/Permissions.cpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "Files.h"
```
[the remark this answers is not in the document]
Comment (a-random-lemurian) [reply source/Permissions.cpp]: Dealt with by 0b79abcc944e2db81fdc12516d4d74401d860577

File: source/Permissions.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#ifndef PERMISSIONS_H_
+#define PERMISSIONS_H_
```
[the remark this answers is not in the document]
Comment (a-random-lemurian) [reply source/Permissions.h]: Dealt with it in commit 780675ff8ebdb36fc41b30e13eee3455de005243

File: source/OutfitterPanel.cpp [hunk 5]
```diff
@@ -694,6 +697,10 @@ void OutfitterPanel::FailSell(bool toStorage) const
 	const string &verb = toStorage ? "uninstall" : "sell";
 	if(!planet || !selectedOutfit)
 		return;
+	else if(!selectedOutfit->CanSell() && !toStorage)
+		GetUI()->Push(new Dialog("You cannot " + verb + " this outfit."));
+	else if(!selectedOutfit->CanUninstall() && toStorage)
+		GetUI()->Push(new Dialog("You cannot " + verb + " this outfit."));
```
Comment (petervdmeer) [inline source/OutfitterPanel.cpp:703]: If an outfit cannot be uninstalled, then selling the ship (the complete ship) should result in the Outfit not being added to the stock (the outfit gets sold with the ship, so it should not be available for separate purchase later).

File: source/OutfitterPanel.cpp [hunk 6]
```diff
@@ -694,6 +697,10 @@ void OutfitterPanel::FailSell(bool toStorage) const
 	const string &verb = toStorage ? "uninstall" : "sell";
 	if(!planet || !selectedOutfit)
 		return;
+	else if(!selectedOutfit->CanSell() && !toStorage)
+		GetUI()->Push(new Dialog("You cannot " + verb + " this outfit."));
+	else if(!selectedOutfit->CanUninstall() && toStorage)
```
Comment (petervdmeer) [inline source/OutfitterPanel.cpp:702]: An outfit that cannot be uninstalled also cannot be sold once it is installed (since selling an installed outfit automatically uninstalls it), so I guess the `toStorage` check here is not needed.

Comment (a-random-lemurian) [reply source/OutfitterPanel.cpp:702]: Which line are we talking about? The check on 700, or 702?

Comment (petervdmeer) [reply source/OutfitterPanel.cpp:702]: The check on line 702; if the outfit cannot be uninstalled, then it doesn't matter if the player tries to put it to storage or if the player tries to sell it; both cases should cause an not-allowed error-message to pop-up.

File: source/Permissions.cpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
+/* Permissions.cpp
+Copyright (c) 2022 by Lemuria
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/Permissions.cpp:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```
The standard copyright header was changed recently.

File: source/Permissions.h [hunk 8]
```diff
@@ -0,0 +1,31 @@
+/* Permissions.h
+Copyright (c) 2022 by Lemuria
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/Permissions.h:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

**Review by tibetiroka — CHANGES_REQUESTED** [review]

File: source/Permissions.h [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+#include "DataNode.h"
+
+class Permissions {
+public:
+	bool CanSell() const;
+	bool CanUninstall() const;
+
+public:
+	void Load (const DataNode &node);
```
Comment (warp-core) [inline source/Permissions.h:24]:
```suggestion
	void Load(const DataNode &node);
```

File: source/Permissions.h [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#ifndef ESKY_PERMISSIONS_H_
+#define ESKY_PERMISSIONS_H_
+
+#include "DataNode.h"
```
Comment (warp-core) [inline source/Permissions.h:16]: Since `DataNode` only appears as a reference in a method argument, it can be forward declared. Include the full header in the .cpp file.
```suggestion
class DataNode;
```

File: source/Permissions.cpp [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+*/
+
+#include "Permissions.h"
+#include "DataNode.h"
```
Comment (warp-core) [inline source/Permissions.cpp:14]:
```suggestion

#include "DataNode.h"
```

File: source/Permissions.cpp [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+using namespace std;
+
```
Comment (warp-core) [inline source/Permissions.cpp:17]:
```suggestion



```

File: source/Permissions.cpp [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+void Permissions::Load (const DataNode &node)
```
Comment (warp-core) [inline source/Permissions.cpp:18]:
```suggestion
void Permissions::Load(const DataNode &node)
```

File: source/Outfit.h [hunk 14]
```diff
@@ -14,7 +14,7 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 #define OUTFIT_H_
 
 #include "Weapon.h"
-
+#include "Permissions.h"
 #include "Dictionary.h"
```
Comment (warp-core) [inline source/Outfit.h:18]: Leave the blank line. `Outfit` inherits from `Weapon` so that header file is included first and has an empty line separating it from the other included headers.
```suggestion

#include "Dictionary.h"
#include "Permissions.h"
```

File: source/Outfit.cpp [hunk 15]
```diff
@@ -196,6 +196,10 @@ void Outfit::Load(const DataNode &node)
 				if(isNewLicense(licenses, grand.Token(0)))
 					licenses.push_back(grand.Token(0));
 		}
+		else if(child.Token(0) == "inability" && (child.HasChildren() || child.Size() >= 2))
```
Comment (warp-core) [inline source/Outfit.cpp:199]: Not sure `inability` is an appropriate name for the `permissions` on an outfit (or ship). This should also perhaps support `remove`.

File: source/Permissions.cpp [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+{
+	for(const DataNode &child : node)
+	{
+		if(child.Token(0) == "sell")
+			canSell = false;
+		else if(child.Token(0) == "uninstall")
+			canUninstall = false;
```
Comment (warp-core) [inline source/Permissions.cpp:25]: These tokens should require a value afterwards, so, for example, a plugin could remove restrictions.

**Comment by Amazinite** [comment]
Going through PRs older than three months to find PRs that could be closed.

OP requested that I close this.
